File: parser.hs
-- Define the data type for S-expressions (Symbolic Expressions)
data SExp = Atom String | SList [SExp]

-- Define how to show S-expressions as strings
instance Show SExp where
    show (Atom v) = v
    show (SList l) = "(" <> concatMap show l <> ")"

-- Tokenize the input string into a list of tokens
tokenize :: String -> [String]
tokenize input = 
  -- Filter out empty strings and split the input by spaces, adding spaces around parentheses
  filter (not . null) $ words $ concatMap (\c -> if c `elem` "()" then [' ', c, ' '] else [c]) input

-- Parse a list of tokens into an S-expression
parseTokens :: [String] -> SExp
parseTokens tokens = case parseTokens' tokens of
  -- If parsing is successful and there are no remaining tokens, return the S-expression
  (sexp, []) -> sexp
  -- Otherwise, throw an error
  _ -> error "Invalid S-expression"

-- Helper function for parsing tokens, returns an S-expression and the remaining tokens
parseTokens' :: [String] -> (SExp, [String])
parseTokens' [] = error "Unexpected end of input"
parseTokens' (token:tokens) =
  case token of
    -- If the token is an opening parenthesis, parse a list
    "(" -> do 
          let (children, rest) = parseList tokens 
          (SList children, rest)
    -- If the token is a closing parenthesis, throw an error
    ")" -> error "Unmatched )"
    -- Otherwise, the token is an atom
    _   -> (Atom token, tokens)

-- Parse a list of tokens into a list of S-expressions
parseList :: [String] -> ([SExp], [String])
parseList tokens = case tokens of
  -- If the list is empty, throw an error
  [] -> error "Unmatched ("
  -- If the next token is a closing parenthesis, return an empty list and the remaining tokens
  ")":rest -> ([], rest)
  -- Otherwise, parse an S-expression and recursively parse the rest of the list
  _ -> do 
        let (sexp, rest) = parseTokens' tokens
        case parseList rest of
            (sexps, finalRest) -> (sexp:sexps, finalRest)

-- Parse an input string into an S-expression
parse :: String -> SExp
parse input = parseTokens (tokenize input)

-- Main function to test the parser
main :: IO ()
main = print $ parse "(+ 1 (* 2 3))"

File: parser.ts
// Define the data type for S-expressions (Symbolic Expressions)
interface SExp {
    type: string;
    value: any;
    toString(): string;
}

function createAtom(value: string): SExp {
    return {
        type: 'atom',
        value: value,
        toString: () => value,
    };
}

function createList(children: SExp[]): SExp {
    return {
        type: 'list',
        value: children,
        toString: () =>
            '(' + children.map((sexp) => sexp.toString()).join('') + ')',
    };
}

// Tokenize the input string into a list of tokens
function tokenize(input: string): string[] {
    return input
        .replace('(', ' ( ')
        .replace(')', ' ) ')
        .split(' ')
        .filter((token) => token.length > 0);
}

// Parse a list of tokens into an S-expression
function parseTokens(tokens: string[]): SExp {
    if (tokens.length === 0) {
        throw new Error('Unexpected end of input');
    }

    const token = tokens.shift();
    if (token === '(') {
        const children: SExp[] = [];
        while (tokens.length > 0 && tokens[0] !== ')') {
            children.push(parseTokens(tokens));
        }
        if (tokens.length === 0) {
            throw new Error('Unmatched (');
        }
        tokens.shift(); // Remove ')'
        return createList(children);
    } else if (token === ')') {
        throw new Error('Unmatched )');
    } else if (typeof token === 'string') {
        return createAtom(token);
    } else {
        throw new Error('Invalid token type');
    }
}

// Parse an input string into an S-expression
function parse(input: string): SExp {
    return parseTokens(tokenize(input));
}

// Main function to test the parser
const input = '(+ 1 (* 2 3))';
const result = parse(input);
console.log(result.toString());
